### PR TITLE
fix: output YAML the way `prettier` likes it

### LINKED ISSUE #4119: Write `dbt` YAML files the way `prettier` likes them
# Overview

When updating some of the `dbt` YAMl files using the `dbt_helper.py` script I noticed that the YAML formatting which the script uses is different from what is enforced by `prettier` which we use to ensure standardized formatting in all of our YAML files via a `pre-commit` hook. This means that every update results in a large number of meaningless diffs, and makes it hard to see what actually changed.

Can we update the script to use the same indentation / formatting so there isn't guaranteed to be a diff that pre-commit has to fix, even if there haven't been any changes to the schema? It might also be good to use sorted keys to avoid diffs from just re-ordering? This might mean using another YAML library beyond the `pyyaml` one we're apparently using now.?

## Success Criteria

* [x] Using the `devtools/dbt_helper.py` script to update table schemas or tests doesn't automatically result in the `prettier` pre-commit hook needing to fix things.
* [ ] Diffs in YAML files are minimized, making it easy to identify only the meaningful changes before the pre-commit hooks run.

### PR DESCRIPTION
**Pull request by jdangerx**
# Overview

Closes #4119.

## What problem does this address?

The YAML files dumped out by `dbt_helper update-tables` didn't indent the lists the same way prettier wants. This adds a weird little custom class that fixes that.

## What did you change?

- change YAML output format
- update `dbt_helper update-tables` so that it runs without errors
- change `dbt_helper update-tables --update` output to have more readable diff

## Documentation

Make sure to update relevant aspects of the documentation:

- [x] Update the [release notes](https://catalystcoop-pudl.readthedocs.io/en/nightly/release_notes.html): reference the PR and related issues.

# Testing

Tried updating a schema and the diff was concise. Also the `prettier` check passed.

```fish
daz@inkwell ~/w/pudl (4119-prettier-fix) [1]> dbt_helper update-tables --schema out_eia__yearly_plants --clobber                                                                               (pudl-dev) 
2025-07-10 16:57:16 [    INFO] catalystcoop./home/daz/work/pudl/src/pudl/scripts/dbt_helper.py:388 Wrote schema config for table out_eia__yearly_plants at dbt/models/output/out_eia__yearly_plants/schema.yml.
daz@inkwell ~/w/pudl (4119-prettier-fix)> git diff dbt                                                                                                                                         (pudl-dev) 
diff --git a/dbt/models/output/out_eia__yearly_plants/schema.yml b/dbt/models/output/out_eia__yearly_plants/schema.yml
index 8c15afada..1b02f2d30 100644
--- a/dbt/models/output/out_eia__yearly_plants/schema.yml
+++ b/dbt/models/output/out_eia__yearly_plants/schema.yml
@@ -7,10 +7,6 @@ sources:
           - check_row_counts_per_partition:
               table_name: out_eia__yearly_plants
               partition_column: report_date
-          - expect_date_frequency_ratio:
-              compare_model: source("pudl", "out_eia__yearly_generators")
-              multiplier: 1
-              date_column: report_date
         columns:
           - name: plant_id_eia
           - name: plant_name_eia
daz@inkwell ~/w/pudl (4119-prettier-fix)> git add dbt                                                                                                                                          (pudl-dev) 
daz@inkwell ~/w/pudl (4119-prettier-fix)> pre-commit run prettier                                                                                                                              (pudl-dev) 
prettier.................................................................Passed
```

Also, tried updating a schema with just `--update` and got a nice diff output.

## To-do list

- [x] Review the PR yourself and call out any questions or issues you have.

### CODE REVIEW

File: src/pudl/logging_helpers.py [hunk 1]
```diff
@@ -22,6 +22,8 @@ def configure_root_logger(logfile: str | None = None, loglevel: str = "INFO"):
     numba_logger.setLevel(logging.WARNING)
 
     logger = logging.getLogger("catalystcoop")
+    logger.propagate = False
```
Comment (jdangerx) [inline src/pudl/logging_helpers.py]: This being `True` was causing some irritating double-logging when I ran `dbt_helper` commands.

File: src/pudl/scripts/dbt_helper.py [hunk 2]
```diff
@@ -194,38 +228,34 @@ def schema_has_removals_or_modifications(diff: DeepDiff) -> bool:
...
-        if isinstance(changes, dict):
-            for path, value in changes.items():
-                summary_elements.append(f"  - {path}: {value}")
-        else:
-            for item in changes:
-                summary_elements.append(f"  - {item}")
 
-    return "\n".join(summary_elements)
+def _schema_diff_summary(old_schema: DbtSchema, new_schema: DbtSchema):
+    """Return a summary of schema changes based on YAML output."""
```
Comment (jdangerx) [inline src/pudl/scripts/dbt_helper.py:247]: The `deepdiff` output is useful for figuring out what types of changes there are (add, remove, etc) but the string output is tough to read. Also, dumping out two whole YAML files is hard to compare. But we can just compare the YAML files with `difflib` and print out something actually kind of nice.

**Comment by jdangerx** [comment]
@marianneke I noticed in testing this PR out that the update logic basically forces us to manually update all schemas that have custom data tests - probably we would want this to preserve any data tests that already exist? But that seems out of scope for this PR (which already has ballooned to doing two things instead of one).

I guess one issue would be "what if the data test no longer refers to an existing column?" but again that seems out of scope here.

```fish
daz@inkwell ~/w/pudl (4119-prettier-fix)> dbt_helper update-tables --schema out_eia__yearly_plants --update                                                                                    (pudl-dev) 
2025-07-10 17:17:13 [ WARNING] catalystcoop./home/daz/work/pudl/src/pudl/scripts/dbt_helper.py:388 
⚠️ WARNING: Some elements would be deleted by this update! Please update manually instead.
2025-07-10 17:17:13 [    INFO] catalystcoop./home/daz/work/pudl/src/pudl/scripts/dbt_helper.py:243 
--- old_schema
+++ new_schema
@@ -7,10 +7,6 @@
           - check_row_counts_per_partition:
               table_name: out_eia__yearly_plants
               partition_column: report_date
-          - expect_date_frequency_ratio:
-              compare_model: source("pudl", "out_eia__yearly_generators")
-              multiplier: 1
-              date_column: report_date
         columns:
           - name: plant_id_eia
           - name: plant_name_eia
2025-07-10 17:17:13 [   ERROR] catalystcoop./home/daz/work/pudl/src/pudl/scripts/dbt_helper.py:410 DBT configuration for table out_eia__yearly_plants has information that would be deleted. Update manually or run with clobber.
```

File: src/pudl/logging_helpers.py (hunk 1 above)
Comment (marianneke) [reply src/pudl/logging_helpers.py]: Could you comment on the change in behavior this will cause? Looking into it briefly, it seems that setting it to `False` is helpful if there is some logging you want to treat in an isolated manner, but could this negatively affect logging behavior in other parts of the code?

Also, would it make more sense to set this to False in the dbt helper module or maybe in all scripts by default (since it makes more sense for that to be isolated)?

File: test/unit/dbt_helper_test.py [hunk 3]
```diff
@@ -612,34 +611,33 @@ def test_complex_schema_diff_output(capsys):
...
-    assert "'description': 'old'" in output, output
-    assert "'description': 'new'" in output, output
-
-    # model2 addition
-    assert "model2" in output, output
-    assert "'description': 'added'" in output, output
+    output = _schema_diff_summary(old_schema, new_schema)
+    expected = """--- old_schema
++++ new_schema
+@@ -1,12 +1,20 @@
```
Comment (marianneke) [inline test/unit/dbt_helper_test.py:617]: I like having an explicit form of the expected output 👍

**Comment by jdangerx** [comment]
> @marianneke I noticed in testing this PR out that the update logic basically forces us to manually update all schemas that have custom data tests - probably we would want this to preserve any data tests that already exist? But that seems out of scope for this PR (which already has ballooned to doing two things instead of one).
> 
> I guess one issue would be "what if the data test no longer refers to an existing column?" but again that seems out of scope here.

One workaround I guess is to hit that `--clobber` and then use `git add -p` to keep the stuff you want from the old version.

File: src/pudl/logging_helpers.py (hunk 1 above)
Comment (jdangerx) [reply src/pudl/logging_helpers.py]: Good question. I tried running `make pytest-coverage` and the integration test logs were a little funky. I decided that instead of chasing that down and figuring out what exactly was going on I should:

* use `click.secho()` for the colored output - that's the "intended" tool for CLI program output that I just learned about from #4163 
* slap that `logger.parent.propagate = False` in the dbt_helper script only

Then I think we don't have to worry about breaking stuff outside of this script when we merge :)

**Comment by marianneke** [comment]
> @marianneke I noticed in testing this PR out that the update logic basically forces us to manually update all schemas that have custom data tests - probably we would want this to preserve any data tests that already exist? But that seems out of scope for this PR (which already has ballooned to doing two things instead of one).
> 
> I guess one issue would be "what if the data test no longer refers to an existing column?" but again that seems out of scope here.

Agree. More elegant automated handling of updates would definitely be helpful. That can get a little convoluted with edge cases though, especially as you said if columns are renamed but also if parameters in the test are changed, and we need to decide whether to just add tests to the column or replace.

File: src/pudl/scripts/dbt_helper.py [hunk 4]
```diff
@@ -20,10 +21,50 @@
...
 
+def _prettier_yaml_dumps(yaml_contents: dict[str, Any]) -> str:
+    """Dump YAML to string that Prettier likes."""
+
+    class PrettierCompatibleDumper(yaml.Dumper):
+        """Custom Dumper that indents lists like prettier does.
+
+        Default dumper behavior::
+
+            foo:
```
Comment (marianneke) [inline src/pudl/scripts/dbt_helper.py:38]: Wait, was the old behavior in DbtSchema.to_yaml` like this? Because then I would have expected to see errors due to incorrect alignment.

**Comment by marianneke** [comment]
Adding note here too for posterity:
The `_prettier_yaml_dumps` functionality you added is quite general so I wonder if that should be moved out of the script and into the core library instead.

File: src/pudl/scripts/dbt_helper.py (hunk 4 above)
Comment (jdangerx) [reply src/pudl/scripts/dbt_helper.py:38]: Yes, this was the old behavior - it's "valid YAML" but also `prettier` runs in our pre-commit hooks which fixes all the alignment issues.
